### PR TITLE
Table of contents

### PR DESCRIPTION
**Pull request by Mihaiii**
Added a table of contents page (`/content/toc`).
https://github.com/StarcounterApps/Simplified/pull/69 is needed for this to work.
There is a "Go" button on search that doesn't really do anything as the search is done after each key is pressed, but I decided to let it there as it makes the "search" input more visible (will remove it if is needed).

### CODE REVIEW

**Comment by Mihaiii** [comment]
> 2. I'm not sure how should it work, but when I've created some content with path(and URL): /Test/Something It's shown directly under Home node. Shouldn't be like Home -> Test -> something? Isn't that how tree view should work?
image

You can select the "Parent Page" when you create a Content item (there is a checkbox in case you want to add a parent page and then a combo).

> 4. IMPO this elements positioning/layout is pretty confusing. I would rather advice to create something more like this:
image

I'm assuming that by "Title" you are referring to the "Url" field, not sure what you mean by the description. The url is changing when you select a content item btw, but ok, I will include it in the layout.

**Comment by Mihaiii** [comment]
> 3. I highly recommend some simple navigation bar to move between views since there is more than 1 view right now (take a look on Website or Calendar (app-rewritting branch) for an example).

Initially I was thinking in terms of creating different views that would serve as `Pin Uri`s in different cases (so for example a repetition of content items that could be used as a footer - something like in `content/table-of-content` that right now lacks CSS). But I guess I could make a nav bar for multiple views and consider adding some sort of category to group the content items (some content items might be for footer, some for other blending point etc.). 

A think that we should find a formula for a stronger connection with the blending points from different surfaces. How do you see it?

**Comment by Mihaiii** [comment]
```
> 2. I'm not sure how should it work, but when I've created some content with path(and URL): /Test/Something It's shown directly under Home node. Shouldn't be like Home -> Test -> something? Isn't that how tree view should work?
![image](https://user-images.githubusercontent.com/25042256/29916949-a3ccd3a4-8e40-11e7-8924-23bc011b2ac9.png)

You can select the "Parent Page" when you create a Content item (there is a checkbox in case you want to add a parent page and then a combo).

```

But anyway, your approach is interesting and I didn't consider it (having `/foo` the default parent for `/foo/bar`). I think I'll need https://github.com/StarcounterApps/Website/issues/54 in order to do it, but I'll implement it.

**Comment by Mihaiii** [comment]
I think that I covered everything other than the nav bar. 

With the nav bar there are some issues since I'm trying to change the structure (NavMasterPage should be the CurrentPage of MasterPage), but I'm hitting all sort of errors.

**Comment by PatrykSzwer** [comment]
I dont understand, why dont You just add nav element into masterPage with links to views like It's in other apps?

**Comment by Mihaiii** [comment]
That's not a problem. Right now `MasterPage` and `NavMasterPage` aren't related, but they should as I said in the previous message ("NavMasterPage should be the CurrentPage of MasterPage").
Do you know any example that uses `starcounter-include` with a `partial` that is a page which also uses `starcounter-include`. I'm doing something wrong here, but the code looks ok so I'd like to look at another example and test some stuff within another app.

**Comment by Mihaiii** [comment]
> Do you know any example that uses starcounter-include with a partial that is a page which also uses starcounter-include? I'm doing something wrong here, but the code seems ok so I'd like to have look at another example and test some stuff within another app.

Nevermind, I found an example in KitchenSink. Can we skip the nav bar for the moment since the problem I'm facing might be related with https://github.com/StarcounterApps/KitchenSink/issues/167 ?

**Comment by Mihaiii** [comment]
@PatrykSzwer Please have another look.

**Comment by PatrykSzwer** [comment]
For some reason I can't run Content app now. I'm getting:
https://github.com/Starcounter/level1/wiki/SCERR6066

image

**Comment by Mihaiii** [comment]
I can't reproduce it.If you checkout to an older commit where you know it worked for you before, it still gives the error?

**Comment by PatrykSzwer** [comment]
Nvm, It was some local problem, I fixed it already.

**Comment by PatrykSzwer** [comment]
Missclick with that closing, my bad sorry 😛

**Comment by PatrykSzwer** [comment]
Well, something strange is happening now. I've had 2 old contents created. First my tree was showing them but on tree as a icons only (empty text field) and on content view as "Content not found". I found that You've added Name field, so I've added name to them and tried again. 
image

Now tree looks ok but content is still "not found". And when I'm edditing my content visual tab is completely blank:
reprod

**Comment by Mihaiii** [comment]
Yes, I made some changes regarding the ids. Could you please just create new content items and delete the old ones?

**Comment by Mihaiii** [comment]
Regarding the "content visual tab is completely blank", I'll try to get a scenario, shouldn't be related with the changes I made.

**Comment by Mihaiii** [comment]
Managed to reproduce the blank visual tab. 👍 
Please pull to get latest commit.

**Comment by Mihaiii** [comment]
By the way, I intend to release it by bumping major since the content items made with Content app 1.0.0 won't work with these changes.

Please let me know if I shouldn't.

File: src/Content/Api/PartialHandlers.cs [hunk 1]
```diff
@@ -17,28 +18,78 @@ public void Register()
...
+
                 page.WebContent.Data = content;
+                //TO DO: improve query
+                var wc = Db.SQL<WebContent>("SELECT i FROM Simplified.Ring6.WebContent i WHERE i.Name IS NOT null and i.Key <> ? and (i.ParentPage <> ? or i.ParentPage is null) and i.Deleted <> ? order by i.Name", key, key, true);
+                
+                //move this on the client side after `juicy-select` allows it.
+                foreach (var wc_content in wc)
+                {
+                    CmsItemPage.AllPagesElementJson elem = new CmsItemPage.AllPagesElementJson();
+                    elem.Name = String.Format("{0} ({1})", wc_content.Name, wc_content.Key);
```
Comment (PatrykSzwer) [inline src/Content/Api/PartialHandlers.cs]: please use string interpolation

File: src/Content/Api/PartialHandlers.cs [hunk 2]
```diff
@@ -17,28 +18,78 @@ public void Register()
...
+                    elem.Key = wc_content.Key;
+                    page.AllPages.Add(elem);
+                }
+
+                page.HasParent = String.IsNullOrEmpty(page.WebContent.ParentPage) ? false : true;
```
Comment (PatrykSzwer) [inline src/Content/Api/PartialHandlers.cs]: Please simplify this expression.

File: src/Content/Api/PartialHandlers.cs [hunk 3]
```diff
@@ -17,28 +18,78 @@ public void Register()
...
+                master.RefreshAll = true;
+                master.PartialUrl = url;
+                master.RefreshCurrentPage();
+
+                return master;
+            }, new HandlerOptions() { SelfOnly = true });
+
+            Handle.GET("/content/partials/empty", () =>
+            {
+                return new EmptyPage();
```
Comment (PatrykSzwer) [inline src/Content/Api/PartialHandlers.cs]: You can use lambda expressions in cases like this.

File: src/Content/ViewModels/CmsItemPage.json.cs [hunk 4]
```diff
@@ -11,53 +11,54 @@ partial class CmsItemPage : Json
...
         void Handle(Input.SaveTrigger action)
         {
-            if (NotEmptyUrl(WebContent.Url) && UniqueUrl(WebContent.Url) && CharAccepted(WebContent.Url))
+            var wc = WebContent;
+            if (NotEmptyName(wc.Name))
             {
+                SetUrl(wc);
+                SetPath(wc);
                 this.Transaction.Commit();
                 var ATarget = new DataHelper.AlertTarget[] { DataHelper.AlertTarget.Main };
```
Comment (PatrykSzwer) [inline src/Content/ViewModels/CmsItemPage.json.cs]: Describing type here is redundant, and variable name is incorrect.

File: src/Content/ViewModels/CmsItemPage.json.cs [hunk 5]
```diff
@@ -11,53 +11,54 @@ partial class CmsItemPage : Json
...
+            if (string.IsNullOrEmpty(Name))
             {
-                if (!ValidChars.Contains(UrlChar))
-                {
-                    var ATarget = new DataHelper.AlertTarget[] { DataHelper.AlertTarget.Main, DataHelper.AlertTarget.Url };
-                    this.Alert = SetAlert(DataHelper.AlertType.Danger, ATarget);
-                    this.Message = "Character '" + UrlChar + "' is not accepted.";
-                    return false;
-                }
+                var ATarget = new DataHelper.AlertTarget[] { DataHelper.AlertTarget.Main, DataHelper.AlertTarget.Name };
```
Comment (PatrykSzwer) [inline src/Content/ViewModels/CmsItemPage.json.cs]: Same as above. Please scroll trough entire solution, look for the incorrect variable names and fix them if You can.

File: src/Content/ViewModels/CmsPage.json.cs [hunk 6]
```diff
@@ -21,14 +21,27 @@ protected override void OnData()
             {
                 base.OnData();
                 this.Href = string.Format("/content/cms/item/{0}", this.Data.Key);
```
Comment (PatrykSzwer) [inline src/Content/ViewModels/CmsPage.json.cs]: Please use string interpolation.

File: src/Content/ViewModels/CmsPage.json.cs [hunk 7]
```diff
@@ -21,14 +21,27 @@ protected override void OnData()
...
             }
 
             protected CmsPage ParentPage => this.Parent.Parent as CmsPage;
 
             void Handle(Input.DeleteTrigger action)
             {
                 this.ParentPage.Items.Remove(this);
-                Db.Transact(() => this.Data.Delete());
+                Db.Transact(() => this.Data.Deleted = true);
+                //https://github.com/Starcounter/RebelsLounge/issues/162
```
Comment (PatrykSzwer) [inline src/Content/ViewModels/CmsPage.json.cs]: Are those comments still necessary?

**Comment by Mihaiii** [comment]
Thanks! Please have another look.

**Comment by PatrykSzwer** [comment]
I found out one more thing. Content view is not fitting correctly into window. It's specially easy to spot with website app running:
image

**Comment by Mihaiii** [comment]
I added backwards compatibility.
Regarding the previous message (fitting in the window), I think it's desired to be that way since there aren't a lot of info displayed. That CSS rule is on [`master` branch also](https://github.com/StarcounterApps/Content/blob/682f1071f3372895012393374801328869de6147/src/Content/wwwroot/Content/views/CmsPage.html#L89) and I'd prefer to let it like that for the moment. There might be some serious CSS changes in the future, but I'll treat them in a separate PR.

**Comment by PatrykSzwer** [comment]
Ok, I agree. You can make separate issue for those CSS rules. So I assume It's ready for review?

**Comment by Mihaiii** [comment]
Yes, it's ready. :)

File: src/Content/Api/PartialHandlers.cs [hunk 8]
```diff
@@ -1,55 +1,110 @@
...
 
             Handle.GET("/content/partials/launchermenu", () => new LauncherMenu());
 
             Blender.MapUri("/content/partials/launchermenu", "menu");
             Blender.MapUri("/content/partials/app-name", "app-name");
         }
+
+        public static TreeMasterPage GetTreeMasterFromSession()
+        {
+            TreeMasterPage TreeMaster = Session.Ensure().Store[nameof(TreeMasterPage)] as TreeMasterPage;
```
Comment (PatrykSzwer) [inline src/Content/Api/PartialHandlers.cs]: Variable should be lowercase.

File: src/Content/ViewModels/CmsItemPage.json.cs [hunk 9]
```diff
@@ -11,76 +11,81 @@ partial class CmsItemPage : Json
...
-            if (NotEmptyUrl(WebContent.Url) && UniqueUrl(WebContent.Url) && CharAccepted(WebContent.Url))
+            var wc = WebContent;
+            if (NotEmptyName(wc.Name))
             {
+                SetUrl(wc);
+                SetPath(wc);
                 this.Transaction.Commit();
-                var ATarget = new DataHelper.AlertTarget[] { DataHelper.AlertTarget.Main };
-                this.Alert = SetAlert(DataHelper.AlertType.Warning, ATarget);
+                DataHelper.AlertTarget[] AlertTargets = new DataHelper.AlertTarget[] { DataHelper.AlertTarget.Main };
```
Comment (PatrykSzwer) [inline src/Content/ViewModels/CmsItemPage.json.cs]: Variable should be lowercase, and type after keyword new is redundant. Should be
`DataHelper.AlertTarget[] alertTargets = new[] { DataHelper.AlertTarget.Main };`

File: src/Content/ViewModels/CmsItemPage.json.cs [hunk 10]
```diff
@@ -11,76 +11,81 @@ partial class CmsItemPage : Json
...
+            if (string.IsNullOrEmpty(Name))
             {
-                if (!ValidChars.Contains(UrlChar))
-                {
-                    var ATarget = new DataHelper.AlertTarget[] { DataHelper.AlertTarget.Main, DataHelper.AlertTarget.Url };
-                    this.Alert = SetAlert(DataHelper.AlertType.Danger, ATarget);
-                    this.Message = "Character '" + UrlChar + "' is not accepted.";
-                    return false;
-                }
+                DataHelper.AlertTarget[] AlertTargets = new DataHelper.AlertTarget[] { DataHelper.AlertTarget.Main, DataHelper.AlertTarget.Name };
```
Comment (PatrykSzwer) [inline src/Content/ViewModels/CmsItemPage.json.cs]: Same as above.

File: src/Content/ViewModels/CmsItemPage.json.cs [hunk 11]
```diff
@@ -11,76 +11,81 @@ partial class CmsItemPage : Json
...
+                this.Alert = SetAlert(DataHelper.AlertType.Danger, AlertTargets);
+                this.Message = "Please complete the \"Name\" field.";
+                return false;
             }
             return true;
         }
 
-        private AlertJson SetAlert(DataHelper.AlertType Type, DataHelper.AlertTarget[] ATargets)
+        private AlertJson SetAlert(DataHelper.AlertType Type, DataHelper.AlertTarget[] AlertTargets)
```
Comment (PatrykSzwer) [inline src/Content/ViewModels/CmsItemPage.json.cs]: Method argument should be lowercase.

File: src/Content/wwwroot/Content/views/CmsPage.html [hunk 12]
```diff
@@ -109,6 +127,8 @@ <h3 class="content-dialog__text">Are you sure you want to delete this content it
             template.deleteItem = function (e) {
                 itemToDelete = e.target.item;
                 this.$.dialogContainer.open();
+                template.set("hasChildren", !!itemToDelete.DirectChildren.length);
```
Comment (PatrykSzwer) [inline src/Content/wwwroot/Content/views/CmsPage.html]: Double explanation mark is considered obscure, I don't think we should use it:
https://stackoverflow.com/a/1406618

**Comment by Mihaiii** [comment]
Thanks! I made the changes. Please have another look.

**Comment by PatrykSzwer** [comment]
Is it ready for next review?

**Comment by Mihaiii** [comment]
Not yet. I have to add more features and after that I'll request a final review. :)

**Comment by Mihaiii** [comment]
@PatrykSzwer Please have a look. There are another desired features, but I need to merge this into develop at the stage it is now.
In `/content/toc` you have that "Go" button that before didn't do anything as the filtering was applied (and still is as you type). Now it displays results if Search app is running.

**Comment by PatrykSzwer** [comment]
I'm still having same issues that I mentioned here https://github.com/StarcounterApps/Content/pull/102#pullrequestreview-61517166 but since It is doesn't happening with newly created ones, lets skip this for now.
About searching, could You explain more, or provide some steps to see the results, about that Go button and search app? Beacuse when I'm using it I can't see any reaction. I have Search app, Website app and SampleWebsiteTheme app (for showing search bar) running.
And second thing, when I 2 contents, one parent and one child, named lets say AAA(p) and BBB(c). When I start typing name of a parent (AA), child that doesn't have those characters in name will be shown, but not anymore in tree, as a child but as a separated content. You have to decide here, either You want to show entire parent tree with proper dependencies showed, or You're don't want to show child of searched content.

**Comment by Mihaiii** [comment]
> I'm still having same issues that I mentioned here #102 (review) but since It is doesn't happening with newly created ones, lets skip this for now.

Mmm, could you please make a quick gif? I was able to reproduce it when you first mentioned it, but now it's suppose to be solved (can't reproduce it anymore).

> About searching, could You explain more, or provide some steps to see the results, about that Go button and search app? Beacuse when I'm using it I can't see any reaction. I have Search app, Website app and SampleWebsiteTheme app (for showing search bar) running.

Right now I'm getting an error at search with the `Simplified` that is in the folder that is created at install.

> Starcounter.DbException: ScErrSchemaCodeMismatch (SCERR4177): Operation failed because input does not match schema. CLR Class is not loaded for table Simplified.Ring6.PersonEvent
> Version: 2.3.1.7314.

It's an error that affects multiple apps and that will be solved soon.
Please skip this search part for the moment - I will come back with a message after it is fixed.

But anyway, I forgot to mention that for this PR you need to
1. Run Search app
2. Go to `search/settings`
3. Delete all rows.
4. Add new entry with  "Concept / Relation" `Simplified.Ring6.WebContent`
5. Rerun Search app

I will also make a gif after that error is fixed (it's outside of Content app).

> And second thing, when I 2 contents, one parent and one child, named lets say AAA(p) and BBB(c). When I start typing name of a parent (AA), child that doesn't have those characters in name will be shown, but not anymore in tree, as a child but as a separated content

That's a bug unless BBB has the searched text ("AA") in the value of "Html". That filter searched within both "Name" and "Html". Could you please check as I couldn't reproduce (I have "AAA" and "BBB" - none of them have anything in "Html" - and when I filter for "aa", only "AAA" is displayed).

**Comment by PatrykSzwer** [comment]
> Mmm, could you please make a quick gif? I was able to reproduce it when you first mentioned it, but now it's suppose to be solved (can't reproduce it anymore).

Sure, here:
rep

> I forgot to mention that for this PR you need to...

I've done all steps You provided and getting exception when using search. Look (to get that last error page I'm just refreshing page): 
repcontent

Search app settings:
image

About that AA and BB, ya You're right, that It's because HTML contained those strings. It was pretty unexpected to me, that It will also search in HTML.

**Comment by Mihaiii** [comment]
I tried it in different ways, but can't reproduce any of the 2 problems with SC 2.3.1.7356 and running Content, SignIn, Website.

I will try more, but if you think there is any particular scenario that might help, please let me know.

**Comment by PatrykSzwer** [comment]
I've updated SC to the newest version, and exception during searching dissipated. Although I'm still not getting same result as You:
repcontent

just nothing happen.

**Comment by PatrykSzwer** [comment]
I will show You my contents in database, maybe It will help You with reproduction.
image

First one with big HTML is newly created, second one under is old one and this one is reproducing issue.

**Comment by Mihaiii** [comment]
I only get that behavior when I don't have Search app running (now I see that I haven't mentioned it in the previous message) or when it's running, but I don't have `Simplified.Ring6.WebContent` in the settings. 

I will test some more.

**Comment by PatrykSzwer** [comment]
If You will not find out anything today, tomorrow we will sit on a call and we will follow step by step reproduction on my machine. This should speed up working on this feature.

**Comment by PatrykSzwer** [comment]
After call with @Mihaiii we found out couple of things:

1. Bug, that when You click Go! button twice (sometimes once) searching stops working.
2. Bug, that when You click Home icon in Table of content view, content preview after clicking on one of the contents stops working.
3. Problem described here https://github.com/StarcounterApps/Content/pull/102#pullrequestreview-61517166 is caused by missing <p></p> tags in content HTML.
We wasn't able to figure out what's the cause of Search app usage problem.

**Comment by Mihaiii** [comment]
2 and 3 should be fixed. The error at 3 was cause by TinyMCE - it happens when the "Html" that is given doesn't start with a tag.

Number 1 might be related with the general "Search app usage problem" (I can't reproduce it).

And related with the blending with the Search app problem, @habib535 helped me by trying to check on his machine to see if he can reproduce it and everything is fine in his case also (Search app gets called). I would like to find out what caused the problem in your case as soon as possible because I need to do a release. Maybe we can talk again and look into the problem?

**Comment by PatrykSzwer** [comment]
As we discovered issue number 1 is related to newest SC version. I've tested it with version 2.3.1.7204 and It works.

**Comment by Mihaiii** [comment]
From what I understand from @habib535 , it might be related with https://github.com/Starcounter/RebelsLounge/issues/126 .

I'll come back and test it after the fix in SC is provided. I'm not sure about the status of https://github.com/Starcounter/RebelsLounge/issues/126 at this moment.

File: src/Content/Api/ContentHandlers.cs [hunk 13]
```diff
@@ -37,24 +39,58 @@ public void Register()
...
 
-        private ContentPage GetContentPage()
-        {
-            if (Session.Current != null && Session.Current.Store[nameof(ContentPage)] is ContentPage)
+            Handle.GET("/content/search/dynamic/{?}", (Request req, string query) =>
             {
-                return Session.Current.Store[nameof(ContentPage)] as ContentPage;
-            }
+                Session.Ensure();
+                string url = req.Uri.ToLower().Replace(DataHelper.UrlPrefix, string.Empty);
```
Comment (PatrykSzwer) [inline src/Content/Api/ContentHandlers.cs]: This variable is not used anywhere, remove it.

File: src/Content/Api/ContentHandlers.cs [hunk 14]
```diff
@@ -37,24 +39,58 @@ public void Register()
...
+                if (!string.IsNullOrEmpty(content.Html))
+                {
+                    Regex rgx = new Regex("<([A-Za-z][A-Za-z0-9]*)\\b[^>]*>(.*?)</\\1>", RegexOptions.Singleline);
+                    var textOnlyMatch = rgx.Matches(content.Html)[0];
+                    string textOnly = textOnlyMatch.Value.Replace("\n", " ");
+                    string partialText = textOnly;
+                    if (partialText.Length > 250)
+                    {
+                        partialText = textOnly.Substring(0, 250);
+                        var lastSpace = partialText.LastIndexOf(" ");
```
Comment (PatrykSzwer) [inline src/Content/Api/ContentHandlers.cs]: Should be `var lastSpace = partialText.LastIndexOf(" ", StringComparison.Ordinal);`

File: src/Content/Api/PartialHandlers.cs [hunk 15]
```diff
@@ -1,55 +1,129 @@
...
+                return master;
+            }, new HandlerOptions() { SelfOnly = true });
+
+            Handle.GET("/content/partials/empty", () => new EmptyPage());
+
+            Handle.GET("/content/partials/deny", () => new DenyPage());
 
             Handle.GET("/content/partials/launchermenu", () => new LauncherMenu());
+            
+            Blender.MapUri<WebContent>("/content/search/dynamic/{?}", new string[] { "searchitem" });
```
Comment (PatrykSzwer) [inline src/Content/Api/PartialHandlers.cs]: You don't have to write `new string[]`, It's redundant. Just go with `new[]`

**Comment by Mihaiii** [comment]
I made the changes.

**Pull request merged**